crates/re_ui/src/list_item2/scope.rs

```diff
@@ -0,0 +1,137 @@
+            let stack: &mut StateStack = writer.get_temp_mut_or_default(*STATE_STACK_ID);
+            stack.0.pop()
+        })
+    }
+
+    /// Returns the current [`State`] to be used by `[ListItem]`s.
+    ///
+    /// For ergonomic reasons, this function will fail by returning a default state if the stack is
+    /// empty. This is an error condition that should be addressed by wrapping `ListItem` code in a
+    /// [`super::list_item_scope`].
+    pub(crate) fn top(ctx: &egui::Context) -> State {
+        ctx.data_mut(|writer| {
+            let stack: &mut StateStack = writer.get_temp_mut_or_default(*STATE_STACK_ID);
+            let state = stack.0.last();
+            if state.is_none() {
+                re_log::warn_once!(
+                    "Attempted to access empty ListItem state stack, returning default state. \
+                    Wrap in a `list_item_scope`."
+                );
+            }
+            state.cloned().unwrap_or_default()
+        })
+    }
+
+    fn peek(ctx: &egui::Context) -> Option<&State> {
+        ctx.data(|reader| reader.get_temp(*STATE_STACK_ID))
+    }
+}
+
+/// Create a scope in which `[ListItem]`s can be created.
+///
+/// This scope serves two purposes:
+/// - Manage the state that is saved across frame (e.g. for tracking column boundary position).
+/// - Manage the range of X coordinates defining the boundaries of the hover/selection highlight.
+///
+/// State is loaded against the provided `id`, and pushed to a global stack, such that calls to this
+/// function may be nested. `ListItem` code will always use the top of the stack as current state.
+///
+/// The hover/selection highlight coordinate range is determined with the following heuristics:
+/// - Value passed as argument if not `None`.
+/// - Value from the parent scope if the scope is nested.
+/// - Clip rectangle's `x_range` (legacy behavior).
+///
+/// Given the above, `list_item_scope` can be used for two potentially distinct use-cases:
+/// 1) Store a suitable `background_x_range` value for use by nested `ListItem`s. This happens,
+///    e.g., close to the top of the `egui::SidePanel::show` closure, where the panel size
+///    information is readily available (e.g. `ui.max_rect().x_range()`).
+/// 2) Limit state sharing for a subgroup of `ListItem`s. This makes it possible to independently
+///    align the columns of two `ListItem`s subgroups, for which a single, global alignment would
+///    be detrimental. This may happen in deeply nested UI code.
+///
+/// TODO(#6156): the background X range stuff is to be split off and generalised for all full-span
+/// widgets.
+pub fn list_item_scope<R>(
+    ui: &mut egui::Ui,
+    id: impl Into<egui::Id>,
+    background_x_range: Option<egui::Rangef>,
+    content: impl FnOnce(&mut egui::Ui) -> R,
+) -> R {
+    /*
+    data contains two set of things:
+    - some per container state
+    - a global state stack that is read by actual list items
+     */
+
+    let id = id.into();
```

Comment:
PR4 changes this into:
```rust
let id = ui.id().with(id.into());
```
This way, the `id` argument only needs to be unique to the "local" ui.

Reply:
(i still manage to confuse myself with these sometime-nested id stuff 😓)

Reply:
> This way, the `id` argument only needs to be unique to the "local" ui.

Make sure that's documented in the docstring of the function